include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp

```diff
@@ -0,0 +1,750 @@
+#pragma once
+
+#include <iostream>
+#include <sstream>
+
+#include "device.hpp"
+#include "device_gemm_multiple_d.hpp"
+#include "common_header.hpp"
+#include "tensor_layout.hpp"
+#include "tensor_descriptor.hpp"
+#include "tensor_descriptor_helper.hpp"
+#include "gridwise_gemm_multiple_d_xdl_cshuffle.hpp"
+#include "gemm_specialization.hpp"
+#include "device_prop.hpp"
+
+namespace ck {
+
+template <typename GridwiseGemm,
+          typename FloatAB,
+          typename FloatDsPointer,
+          typename FloatE,
+          typename AElementwiseOperation,
+          typename BElementwiseOperation,
+          typename CDEElementwiseOperation,
+          typename AGridDesc_AK0_M_AK1,
+          typename BGridDesc_BK0_N_BK1,
+          typename DsGridDescriptor_MBlock_MPerBlock_NBlock_NPerBlock,
+          typename EGridDescriptor_MBlock_MPerBlock_NBlock_NPerBlock,
+          typename Block2ETileMap,
+          bool HasMainKBlockLoop>
+__global__ void
+#if CK_USE_LAUNCH_BOUNDS
+    __launch_bounds__(CK_MAX_THREAD_PER_BLOCK, CK_MIN_BLOCK_PER_CU)
+#endif
+        kernel_gemm_multiple_d_xdl_cshuffle(const FloatAB* __restrict__ p_a_grid,
+                                            const FloatAB* __restrict__ p_b_grid,
+                                            FloatDsPointer p_ds_grid,
```

Comment:
@rocking5566 and I were discussing whether losing `__restrict__` qualifier would require the compiler to apply strict aliasing rule which would hurt performance a bit. We don't have an answer yet but something to take note of nevertheless.

Reply:
you mean if we remove ```__restrict__```, compiler may not do certain optimization, right?

Just curious, is there any benefit in removing this keyword?

Reply:
Removing `__restrict__` means 2 pointers of some type `T` (where the strict aliasing rule applies) could alias each other so kernel would re-read the data from one `T` array after writing to another `T` array.

We can borrow idea from Pytorch, which means to cast pointers to `T* __restrict__` upon `make_dynamic_buffer()`

https://github.com/pytorch/pytorch/blob/73f6601cfcd9b869dde5613ab9c8aec2da28527b/aten/src/ATen/core/TensorAccessor.h#L28-L34

Reply:
>> We can borrow idea from Pytorch, which means to cast pointers to T* __restrict__ upon make_dynamic_buffer()

What is the benefit of doing that?